cupy/_core/_routines_indexing.pyx

```diff
@@ -262,15 +263,31 @@ cpdef list _prepare_slice_list(slices):
                 # keep scalar int
                 continue
 
-        if cupy.min_scalar_type(s).char == 'O':
-            raise IndexError(
-                'arrays used as indices must be of integer (or boolean) type')
         try:
             s = core.array(s, dtype=None, copy=None)
-        except ValueError:
-            # "Unsupported dtype"
+        except ValueError as e:
+            # Conversion failed, presumably with "Unsupported dtype".
+
+            # If this is a NumPy array, it should have an unsupportd dtype
+            # raise that as as not "integer or boolean" dtype.
+            if isinstance(s, numpy.ndarray):
+                if s.dtype.char not in _dtype.all_type_chars:
+                    raise IndexError(
+                        'arrays used as indices must be of integer '
+                        '(or boolean) type')
+                # Unlikely/impossible, but there was another error re-raise
+                raise
+
+            try:
+                numpy.asarray(s)  # check if NumPy can convert the index
+            except Exception:
+                # Can't convert, raise original (probably identical) error.
+                # For example `[1, [2]]` is "ragged" and fails this way.
+                raise e from None
```

Comment:
I admit, this got a bit longer than I hoped for and `min_scalar_type` should convert rarely.

But since the `min_scalar_type` looked unclear to me tried to re-organize it to not try ahead of time.

(For reference it was added in gh-7286.)